docs/decisions/00TBD-function-calling-reliability.md

```diff
@@ -0,0 +1,246 @@
+        {
+            return (pluginName, functionName);
+        }
+
+        return null;
+    }
+
+    // Try to use use hyphen, dot, and underscore sequentially as separators.
+    var result = Parse(context, '-') ??
+                    Parse(context, '.') ??
+                    Parse(context, '_');
+
+    if (result is not null)
+    {
+        return result.Value;
+    }
+
+    // If no separator is found, return the function name as is allowing AI connector to apply default behavior.
+    return (null, context.FunctionFqn);
+}
+```
+
+Pros:
+- It will mitigate but not reduce or completely eliminate function separator hallucinations by applying a custom heuristic specific to the AI model to parse the function FQN.
+- It can be easily implemented in SK AI connectors.
+
+Cons:
+- It can't be implemented in M.E.AI SK adapters because M.E.AI does not have the concept of plugins.
+
+
+Possible implementations:
+```csharp
+// Either at the operation level
+static (string? PluginName, string FunctionName) ParseFunctionFqn(ParseFunctionFqnContext context)
+{
+    ...
+}
+
+FunctionChoiceBehaviorOptions options = new new()
+{
+    FqnParser = ParseFunctionFqn
+};
+
+var settings = new AzureOpenAIPromptExecutionSettings() { FunctionChoiceBehavior = FunctionChoiceBehavior.Auto(options) };
+
+var result = await this._chatCompletionService.GetChatMessageContentAsync(chatHistory, settings, this._kernel);
+
+// Or at the AI connector configuration level
+IKernelBuilder builder = Kernel.CreateBuilder();
+builder.AddOpenAIChatCompletion("<model-id>", "<api-key>", functionNamePolicy: FunctionNamePolicy.Custom("_", ParseFunctionFqn));
+```
+
+### Option 4: Improved Auto-Recovery Mechanism
+
+Currently, when a function that was not advertised is called, SK returns the error message: _"Error: Function call request for a function that wasn't defined."_
+Among the three AI models `gpt-4(0613)`, `gpt-4o-mini(2024-07-18)`, and `gpt-4o(2024-08-06)` only `gpt-4o-mini` can automatically recover from this error and successfully call the function using the correct name. 
+The other two models fail to recover and instead return a final message similar to: _"I'm sorry, but I can't provide the answer right now due to a system error."_
+
+However, by adding function name to the error message - "Error: Function call request for **foo.bar** function that wasn't defined." and 
+the "You can call tools. If a tool call failed, correct yourself." system message to chat history, all three models can auto-recover from the error and call the function using the correct name.
+
+Taking all this into account, we can add function name into the error message and provide recommendations to add the system message to improve the auto-recovery mechanism.
+
+Pros:
+- More models can auto-recover from the error.
+
+Cons:
+- The auto-recovery mechanism may not work for all AI models.
+- M.E.AI will need to update the error message as well.
```

Comment:
We can make whatever changes we want to M.E.AI, but a) the M.E.AI implementation of function calling is already extensible, b) the M.E.AI implementation of function calling can already be completely replaced if someone wants their own thing, and c) SK can provide its own implementation of function calling when using M.E.AI under the covers.